### PR TITLE
Tentative update to recent JuMP/MOI and Julia 1

### PR DESCRIPTION
**Pull request by cynddl**
The exact pruning tests are broken, but the rest is there. Can you please have a look and let me know what to change? Thanks!

### CODE REVIEW

File: Manifest.toml [hunk 1]
```diff
@@ -0,0 +1,249 @@
+# This file is machine-generated - editing it directly is not advised
+
+[[Base64]]
```
Comment (blegat) [inline Manifest.toml]: Add the `Manifest.toml` file in `.gitignore`

**Review by blegat — COMMENTED** [review]
Remove Julia v0.6 and Julia v0.7 from Travis and add Julia v1.3 and Julia v1.4

File: src/exact.jl [hunk 2]
```diff
@@ -68,26 +68,27 @@ function isdominated(A, b, islb, isfun, k, solver, lb, ub, epsilon)
...
-    res = linprog(c, H, -Inf, h, lbx, ubx, solver)
-    if res.status == :Optimal
-        res = res.objval
-        return (islb) ? -res < epsilon : res > -epsilon
-    else
-        return false
-    end
+    # solve the LP with JuMP
+    model = Model(optimizer_constructor)
+    z = @variable(model, [1:nx+1], lower_bound=lbx, upper_bound=ubx)
```
Comment (blegat) [inline src/exact.jl]: Giving infinite bounds is not correctly handled by some solvers.
Create free variables with `@variable(model, [1:nx+1])` and then do
```julia
if ub isa Vector
    set_lower_bound.(z, lb)
end
```
Same for the upper bound

File: Manifest.toml (hunk 1 above)
Comment (cynddl) [reply Manifest.toml]: Good catch, I was always adding this file to repos. Thanks!

Comment (cynddl) [reply Manifest.toml]: Done!

File: src/exact.jl (hunk 2 above)
Comment (cynddl) [reply src/exact.jl]: Done!

File: src/CutPruners.jl [hunk 3]
```diff
@@ -3,12 +3,11 @@
 # A package to manage polyhedral convex functions
 ################################################################################
 
-VERSION < v"0.7.0-beta2.199" && __precompile__()
 
 module CutPruners
 
 using Compat, Compat.LinearAlgebra, Compat.SparseArrays
```
Comment (blegat) [inline src/CutPruners.jl]: Replace by `using LinearAlgebra, SparseArrays, Test`

File: Project.toml [hunk 4]
```diff
@@ -0,0 +1,21 @@
+name = "CutPruners"
+uuid = "7b25cd80-e7d3-5169-9be9-645fec4aeb50"
+repo = "https://github.com/cynddl/CutPruners.jl"
+version = "0.1.3"
+
+[deps]
+Compat = "34da2185-b29b-5c13-b0c7-acf172513d20"
```
Comment (blegat) [inline Project.toml]: Remove Compat, no need for it anymore as we drop Julia v0.6 and v0.7

Comment (frapac) [reply Project.toml]: I would add `LinearAlgebra` and `SparseArrays`  as well

File: src/exact.jl [hunk 5]
```diff
@@ -68,26 +68,34 @@ function isdominated(A, b, islb, isfun, k, solver, lb, ub, epsilon)
...
-    res = linprog(c, H, -Inf, h, lbx, ubx, solver)
-    if res.status == :Optimal
-        res = res.objval
-        return (islb) ? -res < epsilon : res > -epsilon
-    else
-        return false
+    ubx = isa(ub, Vector) ? vcat(Inf, ub) : ub
+
+    # solve the LP with JuMP
+    model = Model(optimizer_constructor)
```
Comment (frapac) [inline src/exact.jl:86]: I understand the need to use `JuMP` here, even if it is a non-trivial dependency. In the long term, I would suggest to have a look on a MOI `linprog` , e.g. as in [this package](https://github.com/joaquimg/MatrixOptInterface.jl).

File: test/runtests.jl [hunk 6]
```diff
@@ -1,5 +1,6 @@
 using CutPruners
-using Compat, Compat.Test
+using Compat
```
Comment (frapac) [inline test/runtests.jl]: I would remove `Compat` there

**Review by frapac — CHANGES_REQUESTED** [review]
Thanks for the update!

File: src/exact.jl (hunk 5 above)
Comment (cynddl) [reply src/exact.jl:86]: I had a brief look at MatrixOptInterface.jl but wasn't sure if it was ready and/or maintained.
Is there another simple replacement for MathProgBase.linprog?

Comment (blegat) [reply src/exact.jl:86]: It's still premature to use MatrixOptInterface. Using MOI directly could have the advantage of not being tied to `Float64`. Jump is ok for now

**Comment by cynddl** [comment]
@blegat @frapac Can you check now? The tests are passing, Compat has been removed.

File: src/exact.jl [hunk 7]
```diff
@@ -97,5 +97,10 @@ function isdominated(A, b, islb, isfun, k, optimizer_constructor, lb, ub, epsilo
     optimize!(model)
 
     res = objective_value(model)
-    return (islb) ? -res < epsilon : res > -epsilon
+
+    if termination_status(model) == JuMP.MOI.OPTIMAL
+        return (islb) ? -res < epsilon : res > -epsilon
+    else
+        return false
```
Comment (blegat) [inline src/exact.jl]: If the terminaton status is `JuMP.MOI.INFEASIBLE` then `return`ing `false` is the right thing to do but in any other case, we should display an error including the raw status (see https://www.juliaopt.org/JuMP.jl/dev/solutions/#JuMP.raw_status) and the termination status

Comment (blegat) [reply src/exact.jl]: To correct my earlier message:
If it's `INFEASIBLE` then it's redundant so we should return `true`.
If it's `DUAL_INFEASIBLE` (i.e. unbounded`) then we should return `false`.

Comment (cynddl) [reply src/exact.jl]: I'm confused.
INFEASIBLE => true
DUAL_INFESAIBLE => false
OPTIMAL => (islb) ? -res < epsilon : res > -epsilon

Correct? What do you mean by displaying an error? Printing or yielding? There's no verbose flag at the moment.

Comment (blegat) [reply src/exact.jl]: Yes
```julia
if status == MOI.INFEASIBLE
    return true
elseif status == MOI.DUAL_INFEASIBLE
    return false
elseif status == MOI.OPTIMAL
    return ...
else
    error("Solver returned status $status: $(raw_status(model).)")
end

Comment (blegat) [reply src/exact.jl]: Bump :)

**Pull request merged**